test/libsolidity/ASTPropertyTest.cpp

```diff
@@ -0,0 +1,224 @@
+    {
+        ++m_position;
+        if (isPastEnd())
+            m_currentChar = 0;
+        else
+            m_currentChar = m_input[m_position];
+    }
+
+    vector<StringPair> KeyValueParser::parse()
+    {
+        reset();
+        vector<StringPair> parsedPairs;
+        string testId;
+        string testValue;
+        while (!isPastEnd())
+        {
+            skipWhitespace();
+            if (isIdentifierStart(m_currentChar))
+            {
+                testId.clear();
+                while(isIdentifierPart(m_currentChar))
+                {
+                    testId += m_currentChar;
+                    advance();
+                }
+
+                skipWhitespace();
+                soltestAssert(m_currentChar == ':');
+                advance();
+                skipWhitespace();
+
+                testValue.clear();
+                while(isValidValue(m_currentChar))
+                {
+                    testValue += m_currentChar;
+                    advance();
+                }
+                parsedPairs.emplace_back(testId, testValue);
+            }
+        }
+        return parsedPairs;
+    }
+}
+
+ASTPropertyTest::ASTPropertyTest(string const& _filename):
+	TestCase(_filename)
+{
+	if (!boost::algorithm::ends_with(_filename, ".sol"))
+		BOOST_THROW_EXCEPTION(runtime_error("Invalid test contract file name: \"" + _filename + "\"."));
+
+    m_source = m_reader.source();
+    readExpectations();
+}
+
+void ASTPropertyTest::generateTestCaseValues(string& _values, bool _obtained)
+{
+    _values.clear();
+    for (auto testId: m_expectationsSequence)
+    {
+        soltestAssert(m_testCases.count(testId) > 0);
+        _values +=
+            testId +
+            ": " +
+            (_obtained ? m_testCases[testId].obtainedValue : m_testCases[testId].expectedValue)
+            + "\n";
+    }
+}
+
+void ASTPropertyTest::readExpectations()
+{
+    KeyValueParser parser{m_reader.simpleExpectations()};
+    for (auto [testId, testExpectation] : parser.parse())
+    {
+        m_testCases.emplace(testId, ASTPropertyTestCase{testId, "", testExpectation, ""});
+        m_expectationsSequence.push_back(testId);
+    }
+    generateTestCaseValues(m_expectation, false);
+}
+
+optional<Json::Value> ASTPropertyTest::findNode(Json::Value const& _root, string_view const& _property)
+{
+    if (!_property.empty())
+    {
+        string subNode = string(_property.substr(0, _property.find_first_of('.')));
+        if (subNode != _property)
+            return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
+        else if (_root.isMember(subNode))
+            return make_optional(_root[subNode]);
+    }
+    return {};
+}
+
+void ASTPropertyTest::readTestedProperties(Json::Value const& _astJson)
+{
+    queue<Json::Value> nodesToVisit;
+    nodesToVisit.push(_astJson);
+    string documentation = "documentation";
+    string testCaseLine;
+
+    while(!nodesToVisit.empty())
```

Comment:
> with nlohman this would have been a single for loop over a ::flattened JSON value :D
That would be nice indeed. 

I had no particular reason for choosing this iterative approach over the recursive one. I didn't see any downside, so I just keep it.